src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs

```diff
@@ -137,16 +137,46 @@ internal override FlowAnalysisAnnotations FlowAnalysisAnnotations
                     // https://github.com/dotnet/roslyn/issues/30078: Make sure this is covered by test
                     return annotations.Value;
                 }
+                return DecodeFlowAnalysisAttributes(GetDecodedWellKnownAttributeData());
+            }
+        }
 
-                ParameterWellKnownAttributeData attributeData = GetDecodedWellKnownAttributeData();
-                bool hasEnsuresNotNull = attributeData?.HasEnsuresNotNullAttribute == true;
-
-                return FlowAnalysisAnnotationsFacts.Create(
-                    notNullWhenTrue: hasEnsuresNotNull || attributeData?.HasNotNullWhenTrueAttribute == true,
-                    notNullWhenFalse: hasEnsuresNotNull || attributeData?.HasNotNullWhenFalseAttribute == true,
-                    assertsTrue: attributeData?.HasAssertsTrueAttribute == true,
-                    assertsFalse: attributeData?.HasAssertsFalseAttribute == true);
+        private static FlowAnalysisAnnotations DecodeFlowAnalysisAttributes(ParameterWellKnownAttributeData attributeData)
+        {
+            if (attributeData == null)
+            {
+                return FlowAnalysisAnnotations.None;
+            }
+            FlowAnalysisAnnotations annotations = FlowAnalysisAnnotations.None;
+            if (attributeData.HasAllowNullAttribute) annotations |= FlowAnalysisAnnotations.AllowNull;
+            if (attributeData.HasDisallowNullAttribute) annotations |= FlowAnalysisAnnotations.DisallowNull;
+            if (attributeData.HasMaybeNullAttribute)
+            {
+                annotations |= FlowAnalysisAnnotations.MaybeNull;
             }
+            else
+            {
+                bool? value = attributeData.MaybeNullWhenAttribute;
+                if (value.HasValue)
+                {
+                    annotations |= (value.GetValueOrDefault() ? FlowAnalysisAnnotations.MaybeNullWhenTrue : FlowAnalysisAnnotations.MaybeNullWhenFalse);
```

Comment:
>FlowAnalysisAnnotations.MaybeNullWhenTrue : FlowAnalysisAnnotations.MaybeNullWhenFalse [](start = 64, length = 86)

It doesn't look like for PEParameterSymbol these bits are set conditionally based on absence of MaybeNullAttribute. Whichever logic we use, it should be consistent across PE and source.

Reply:
The bits are set conditionally in `PEParameterSymbol.DecodeFlowAnalysisAttributes`.

---
In reply to: [286264390](https://github.com/dotnet/roslyn/pull/35846#discussion_r286264390) [](ancestors = 286264390)